### ISSUE TITLE
[IDEA] It is proposed to add the TAB key / character to the editor and toolbar

### ISSUE THREAD

**Issue by pmario** [Jermolene/TiddlyWiki5#7778]
We do get a lot examples posted to Talk-TiddlyWiki. -- Since many of them are created and then copy/pasted from TW-com  they do not contain any indentation, because the default editor does not accept the TAB-key. 

**So it's proposed** to add that key by default. It should be easy for users to create code like this:

```
\whitespace trim
\procedure test()
<$list filter="[tag[HelloThere]]">
	<$text text=<<currentTiddler>>/>
	<br>
</$list>
\end

<<test>>
```

The behaviour when the TAB-button is pressed should be as follows: 

**Update proposal** due to some feedback: https://github.com/Jermolene/TiddlyWiki5/issues/7778#issuecomment-1755178186

<details><summary>Old proposal</summary>

- Insert TAB character at current cursor position
- If any text is selected in 1 line
  - Indent the line
  - if SHIFT key is active de-dent the line
- If any test in multiple lines is selected  
  - Indent **all** lines
  - if SHIFT key is active de-dent **all** lines

</detailsy>

**Comment by Jermolene** [comment]
Hi @pmario I'd want to understand the impact on keyboard users and users of screen readers and other assistive technology. I don't think it should be switched on by default.

**Comment by pmario** [comment]
>I'd want to understand the impact on keyboard users [. . .] 

I think it depends on the expectation of the user. 

- If users expect a web-form they will expect tab to jump to the next input element or a button
- If users expect a text editor they do expect the tab button to work. 

From my point of view TW text input is a text editor. Since tab does only jump to text input fields. We do not jump to any buttons, like web-forms would do. 

>[. . .] and users of screen readers and other assistive technology.

Since **rendered** HTML code does ignore whitespace all assistive tech that know does so too. So there should be no impact at all. 

We use indentation in many of your core UI tiddlers. IMO The problem for assistive tech comes from our somewhat invalid DOM structure and not from whitespace. 

>I don't think it should be switched on by default.

So it should be invisible by default and be configurable like any other editor toolbar button?

**Comment by rmunn** [comment]
If you introduce the Tab key for indentation, you'll run into multiple issues around that, which code editors have been dealing with for years:

- Some people prefer the tab character (U+0009) for indentation, others prefer the space character (U+0020)
- People have different preferences for indentation levels (should a tab character mean 8 spaces? 4 spaces? 2?)
- If you paste code that has tab characters in it into your editor where you have used spaces for indentation, you end up with a mix of tabs and spaces in the same file/tiddler, which can lead to inconsistent indentation later

Basically you're going to want at least three options that advanced users can configure:

- Should the Tab key insert U+0009 (`\t`) or U+0020 (space)?
- If it should insert space, how many spaces per indent level?
- Regardless of whether Tab should insert U+0009 or U+0020, how many spaces should the Tab character be *displayed* as?

Oh, and if you're using spaces for indentation, there's also an expectation that the Backspace key will, if pressed when there's nothing but indentation to the left of the cursor, will delete one whole indent level at once, e.g. pressing Backspace once will delete 4 spaces at once if at the left of the line, but only one space at once if the cursor is past the beginning of the line (i.e., there are any non-space characters left of the cursor). You can see this happening in [the Monaco editor demo](https://microsoft.github.io/monaco-editor/) where the Tab key is inserting space characters, and the Backspace key is deleting a group of 4 space characters at once.

There's a lot that goes into handling the Tab key correctly: it's not nearly as simple as "press Tab, insert one tab character".

**Comment by pmario** [comment]
This proposal should discuss the most basic indentation behaviour to be used in a TiddlyWiki context. So yes -- it is opinionated and @Jermolene already mentioned accessibility concerns.

If every aspect should be controllable as @rmunn described, we do have the codemirror plugin + settings.

But -- I think it's a complete **overkill** to install a 500+kByte plugin or 10MB plugin in case of the Monaco editor just to get indentations going, for a **single page app** like TW.

- Historically TW uses tabs for **file-size** reasons
- There is an other good reason -- instead of consistency -- to use tabs
  - see: [Nobody talks about the real reason to use Tabs over Spaces ](https://www.reddit.com/r/javascript/comments/c8drjo/nobody_talks_about_the_real_reason_to_use_tabs/)
- The default tab size in TW is: 4. It is [defined using an existing CSS setting](https://github.com/Jermolene/TiddlyWiki5/blob/ba5bfd1ad07b8afc54a329f942a93142b39a673c/themes/tiddlywiki/vanilla/reset.tid#L25)
- Users can easily add their own CSS settings to 2, 4, 8 or any other indentation they see fit, according to there accessibility needs

IMO none of the "space" related problems you mentioned apply to a minimum viable editor suggested in the OP using tabs.

---

At the moment our users are forced to manually use spaces for indentation, which is a painpoint as the OP and as @rmunn pointed out.

**Comment by mateuszwilczek** [comment]
@pmario in OP you refer to the default editor not accepting `TAB-key` (as in physical keyboard key) and then in the proposed solution you refer to `TAB-button` (as in clickable button in editor toolbar). I think some of the discussion above – like the default behavior or accessibility (jumping between fields with physical TAB key) – results from mixing up of these terms.

This is how I imagine it could work well for everyone. Perhaps it's exactly what you meant, I just thought it wasn't immediately clear.

`TAB-button` is a clickable button in editor toolbar, with the action described in OP.
In order not to mess with the default (accessibility related?) behavior of jumping between fields with `TAB-key`, there are two solutions:
* `TAB-button` is visible in the editor toolbar by default (user can insert tabs by clicking it), but has no keyboard shortcut bound to it.
* `TAB-button` has the shortcuts keyboard `Tab`/`Shift Tab` bound to it, but it is by default hidden from the editor toolbar (making the keyboard shortcuts inactive).

---

I'd like to propose and discuss a small modification to this idea, please let know what you think.

We could split the functionality of `Tab` and `Shift Tab` into two editor toolbar buttons called something like `Increase indent` and `Decrease indent` (with keyboard shortcuts bound to them being respectively `Tab` and `Shift Tab`). Advantages and disadvantages:
* (+) Easier to implement with the current mechanism of a toolbar button having a single purpose, with all keyboard shortcuts assigned to it acting the same.
* (+) Easier to understand for users used to rich text editors, which usually have two such buttons in the GUI, here are examples from Google Docs ![image](https://github.com/Jermolene/TiddlyWiki5/assets/36714554/2692beae-da0b-492b-ad79-8131089b2684) and Word ![image](https://github.com/Jermolene/TiddlyWiki5/assets/36714554/b1d491fb-628a-42ef-a8a8-a13b3a1ace2e).
* **(+) Usable on mobile devices.**
* (–) Two new toolbar buttons instead of one, but that's hardly a clutter-problem since the toolbar is user-configurable.
* (–) Having the action "Insert TAB character at current cursor position if no text is selected" would collide with making the button `Increase indent` act exactly like in rich text editors (where it simply indents the line where the cursor is). Inserting tab characters with `Increase indent` button isn't ideally intuitive; but neither is the alternative of doing all by (shift)-clicking a single `TAB-button` in my view.

**Comment by pmario** [comment]
- The OP at **Talk TiddlyWiki Forum** was updated with
- The OP here stays and only links to the post now

Updated (23.10.10) Proposal

- There should be 2 new buttons eg: ![273841254-b1d491fb-628a-42ef-a8a8-a13b3a1ace2e|48x27](upload://wJRlew1yqi84CEdTD7Rm90K1ybj.png) with 2 independent functions
  - 1 button for indent, 1 button for dedent
  - 2 buttons are needed for mobile devices
- The "indent" button will increase the indentation by 1 level for the line where the cursor is placed. 
  - If more lines are selected the level will be increased for those lines. 
  - It's not needed to select the whole line to use indent and dedent
- The "dedent" button removes 1 indent level for 1 or more lines
- It should be possible to use TAB and SHIFT-TAB keyboard shortcuts with only 1 button visible.
- The keyboard shortcuts should be configurable from the ControlPanel -> Keyboard Shortcut page

**Comment by Alzacon** [comment]
I agree with @Jermolene , it should be a option, because for user of screen readers it would be a trap. 
And when it is enabled. How can they move out of editor?

**Comment by Jermolene** [comment]
Thank you everyone.

It seems that the requirements here are more complex than appears at first sight.

There is also a danger that the more features that we add to the core from programming editors then the more requests we'll get to go further with more features that are common in programming editors (search and replace, multiple cursors, etc).

Given the reservations about accessibility, I think that the best way to satisfy all those requirements is via the optional CodeMirror plugin.

**Comment by pmario** [comment]
@Jermolene I did close [the thread at Talk](https://talk.tiddlywiki.org/t/poll-add-tab-key-character-to-the-tw-default-editor/8183)

You are right. The requirements are much more complex than it appears at first. **But** IMO codemirror is _not_ a real solution. It needs about 500kByte and CM6 seems to need 1MB+, which is a complete "no go" for me personally.

We do have several much smaller plugins that are very close to what users seem to want. 

There is @saqimtiaz's [AutoList](https://saqimtiaz.github.io/sq-tw/editor-autolists.html) plugin and there is my [custom-markup](https://wikilabs.github.io/editions/custom-markup/) plugin, which implement a very similar functionality with a different focus. 
But non the less the underlaying code seems to be very similar. 

We both use the existing toolbar and shortcut configurations to create the needed buttons. ... But we also need new js-code to make it happen. 

So I think, it would be nice if some new generic code would be part of the core, so the requested funcions could be implemented using wikitext only.

I'll close this issue for now and will probably recreate a completely new one after more research about how the "generic code" could look like.

**Comment by pmario** [comment]
I changed my mind. I'll reopen it -- till the new issue is created. So we do not forget about it ;)
